packages/editor/src/lib/editor/derivations/notVisibleShapes.ts

```diff
@@ -0,0 +1,95 @@
+import { RESET_VALUE, computed, isUninitialized } from '@tldraw/state'
+import { TLPageId, TLShapeId, isShape, isShapeId } from '@tldraw/tlschema'
+import { Box } from '../../primitives/Box'
+import { Editor } from '../Editor'
+
+function isShapeNotVisible(editor: Editor, id: TLShapeId, viewportPageBounds: Box): boolean {
+	const maskedPageBounds = editor.getShapeMaskedPageBounds(id)
+	// if the shape is fully outside of its parent's clipping bounds...
+	if (maskedPageBounds === undefined) return true
+
+	// if the shape is fully outside of the viewport page bounds...
+	return !viewportPageBounds.includes(maskedPageBounds)
+}
+
+/**
+ * Incremental derivation of not visible shapes.
+ * Non visible shapes are shapes outside of the viewport page bounds and shapes outside of parent's clipping bounds.
+ *
+ * @param editor - Instance of the tldraw Editor.
+ * @returns Incremental derivation of non visible shapes.
+ */
+export const notVisibleShapes = (editor: Editor) => {
+	const isCullingOffScreenShapes = Number.isFinite(editor.renderingBoundsMargin)
+	const shapeHistory = editor.store.query.filterHistory('shape')
+	let lastPageId: TLPageId | null = null
+	let prevViewportPageBounds: Box
+
+	function fromScratch(editor: Editor): Set<TLShapeId> {
+		const shapes = editor.getCurrentPageShapeIds()
+		lastPageId = editor.getCurrentPageId()
+		const viewportPageBounds = editor.getViewportPageBounds()
+		prevViewportPageBounds = viewportPageBounds.clone()
+		const notVisibleShapes = new Set<TLShapeId>()
+		shapes.forEach((id) => {
+			if (isShapeNotVisible(editor, id, viewportPageBounds)) {
+				notVisibleShapes.add(id)
+			}
+		})
+		return notVisibleShapes
+	}
```

Comment:
are we still planning to do an incremental rbush to make this step faster?

Reply:
To me it seems plenty fast enough right now!

Reply:
Tested it [here](https://github.com/tldraw/tldraw/pull/3420), seems the current from scratch is quite fast. I switched between two areas, one with 1999 shapes and the other one with 1 shape.

This branch:
![CleanShot 2024-04-09 at 16 45 46](https://github.com/tldraw/tldraw/assets/2523721/475af6f4-994c-4f2f-b7d7-29a90f598d1d)

Rbush for from scratch:

![image](https://github.com/tldraw/tldraw/assets/2523721/73ebbbb1-e4ee-412f-893c-72c0ed1e98fe)

I guess the main benefit would be in the incremental part, but that one seems fast enough.